File: app/ingest/text_norm.py
"""
Text de-obfuscation shared by the content analyzer.

Attackers disguise words so keyword/regex matching misses them:
  1. Homoglyphs — Latin-looking letters from other scripts ("urgеnt" with a
     Cyrillic 'е'). Same glyph to a human, different code point to a computer.
  2. Invisible characters — zero-width / soft-hyphen unicode inserted between
     letters ("ver​ify") to break word boundaries.

`canonical()` reverses both so matching sees the real word; the detectors report
when either technique was used, which is itself a strong evasion signal.
"""

from __future__ import annotations

import re
import unicodedata

# Non-Latin letters that look like Latin ones -> their Latin twin.
_CONFUSABLES = str.maketrans({
    # Cyrillic
    "а": "a", "е": "e", "о": "o", "р": "p", "с": "c", "х": "x", "у": "y",
    "ѕ": "s", "і": "i", "ј": "j", "һ": "h", "к": "k", "м": "m", "т": "t",
    "в": "b", "н": "h", "д": "d", "г": "r", "п": "n", "л": "n",
    # Greek
    "ο": "o", "α": "a", "ε": "e", "ρ": "p", "τ": "t", "ν": "v", "ι": "i",
    "κ": "k", "χ": "x", "υ": "u", "μ": "u",
})

# Zero-width / directional / soft-hyphen characters used to split words.
_INVISIBLE = "​‌‍⁠﻿­‎‏᠎"
_INVIS_RE = re.compile(f"[{_INVISIBLE}]")
# an invisible char sitting *between* two letters — i.e. splitting a word
_INVIS_IN_WORD_RE = re.compile(f"(?<=\\w)[{_INVISIBLE}]+(?=\\w)")


def strip_invisible(text: str) -> str:
    return _INVIS_RE.sub("", text or "")


def canonical(text: str) -> str:
    """NFKC-fold, remove invisibles, map homoglyphs to Latin, lowercase."""
    t = unicodedata.normalize("NFKC", text or "")
    t = _INVIS_RE.sub("", t)
    return t.translate(_CONFUSABLES).lower()


def _script(ch: str) -> str | None:
    if not ch.isalpha():
        return None
    try:
        name = unicodedata.name(ch)
    except ValueError:
        return None
    if "CYRILLIC" in name:
        return "CYRILLIC"
    if "GREEK" in name:
        return "GREEK"
    if "LATIN" in name:
        return "LATIN"
    return "OTHER"


def mixed_script_words(text: str) -> list[str]:
    """Words that mix scripts within a single token (e.g. Latin + Cyrillic)."""
    out = []
    for word in re.findall(r"\S+", text or ""):
        scripts = {s for s in (_script(c) for c in word) if s in ("LATIN", "CYRILLIC", "GREEK")}
        if len(scripts) > 1:
            out.append(word)
    return out


def obfuscation_report(text: str) -> dict:
    """Summarise disguise techniques found in `text`."""
    mixed = mixed_script_words(text)
    invis_runs = len(_INVIS_IN_WORD_RE.findall(text or ""))
    return {
        "mixed_script_words": mixed[:5],
        "mixed_script_count": len(mixed),
        "invisible_runs": invis_runs,
        "obfuscated": bool(mixed) or invis_runs >= 1,
    }

File: web/src/lib/ingest/text_norm.ts
/**
 * Text de-obfuscation shared by the content analyzer
 * (ported from app/ingest/text_norm.py).
 *
 * Attackers disguise words so keyword matching misses them: homoglyphs (Latin-
 * looking letters from other scripts) and invisible characters (zero-width /
 * soft-hyphen) inserted between letters. canonical() reverses both; the detector
 * reports when either technique was used, which is itself a strong evasion signal.
 */

// Non-Latin letters that look like Latin ones -> their Latin twin.
const CONFUSABLES: Record<string, string> = {
  // Cyrillic
  а: 'a', е: 'e', о: 'o', р: 'p', с: 'c', х: 'x', у: 'y',
  ѕ: 's', і: 'i', ј: 'j', һ: 'h', к: 'k', м: 'm', т: 't',
  в: 'b', н: 'h', д: 'd', г: 'r', п: 'n', л: 'n',
  // Greek
  ο: 'o', α: 'a', ε: 'e', ρ: 'p', τ: 't', ν: 'v', ι: 'i',
  κ: 'k', χ: 'x', υ: 'u', μ: 'u',
};

// Zero-width / directional / soft-hyphen / mongolian-vowel-separator.
const INVISIBLE = '​‌‍⁠﻿­‎‏᠎';
const INVIS_RE = new RegExp(`[${INVISIBLE}]`, 'g');
// \w is ASCII-only in JS; use a Unicode letter/number class so invisible chars
// inserted BETWEEN non-Latin letters (Cyrillic/Greek homoglyph words) are still
// counted as an obfuscation run — not just between Latin letters.
const INVIS_IN_WORD_RE = new RegExp(`(?<=[\\p{L}\\p{N}_])[${INVISIBLE}]+(?=[\\p{L}\\p{N}_])`, 'gu');

const RE_CYRILLIC = /\p{Script=Cyrillic}/u;
const RE_GREEK = /\p{Script=Greek}/u;
const RE_LATIN = /\p{Script=Latin}/u;

function fold(s: string): string {
  let out = '';
  for (const ch of s) out += CONFUSABLES[ch] ?? ch;
  return out;
}

export function stripInvisible(text: string): string {
  return (text || '').replace(INVIS_RE, '');
}

/** NFKC-fold, remove invisibles, map homoglyphs to Latin, lowercase. */
export function canonical(text: string): string {
  const t = (text || '').normalize('NFKC').replace(INVIS_RE, '');
  return fold(t).toLowerCase();
}

function scriptOf(ch: string): 'CYRILLIC' | 'GREEK' | 'LATIN' | 'OTHER' | null {
  if (!/\p{L}/u.test(ch)) return null;
  if (RE_CYRILLIC.test(ch)) return 'CYRILLIC';
  if (RE_GREEK.test(ch)) return 'GREEK';
  if (RE_LATIN.test(ch)) return 'LATIN';
  return 'OTHER';
}

/** Words that mix scripts within a single token (e.g. Latin + Cyrillic). */
export function mixedScriptWords(text: string): string[] {
  const out: string[] = [];
  for (const word of (text || '').split(/\s+/).filter(Boolean)) {
    const scripts = new Set<string>();
    for (const c of word) {
      const s = scriptOf(c);
      if (s === 'LATIN' || s === 'CYRILLIC' || s === 'GREEK') scripts.add(s);
    }
    if (scripts.size > 1) out.push(word);
  }
  return out;
}

export interface ObfuscationReport {
  mixed_script_words: string[];
  mixed_script_count: number;
  invisible_runs: number;
  obfuscated: boolean;
}

export function obfuscationReport(text: string): ObfuscationReport {
  const mixed = mixedScriptWords(text);
  const invisRuns = ((text || '').match(INVIS_IN_WORD_RE) || []).length;
  return {
    mixed_script_words: mixed.slice(0, 5),
    mixed_script_count: mixed.length,
    invisible_runs: invisRuns,
    obfuscated: mixed.length > 0 || invisRuns >= 1,
  };
}
